libvast/vast/system/importer.hpp

```diff
@@ -29,27 +29,86 @@
 namespace vast::system {
 
 /// Receives chunks from SOURCEs, imbues them with an ID, and relays them to
-/// ARCHIVE and INDEX.
+/// ARCHIVE, INDEX and continuous queries.
 struct importer_state {
+  /// A simple generator for IDs.
+  struct id_generator {
+    /// The next available ID.
+    id i;
+
+    /// The first unavailable ID.
+    id last;
+
+    inline id_generator(id from, id to) : i(from), last(to) {
+      // nop
+    }
+
+    /// Returns whether this generator is exhausted.
+    inline bool at_end() const noexcept {
+      return i == last;
+    }
+
+    /// Returns the next ID and advances the position in the range.
+    inline id next() noexcept {
+      return i++;
+    }
+
+    /// Returns how many more times `next` returns a valid ID.
+    inline int32_t remaining() const noexcept {
+      return static_cast<int32_t>(last - i);
+    }
+  };
+
+  /// Handle to the meta store for obtaining more IDs.
   meta_store_type meta_store;
+
+  /// Handle to the ARCHIVE for persisting incoming events.
   caf::actor archive;
+
+  /// Handle to the INDEX for forwarding incoming events.
   caf::actor index;
-  id next = 0;
-  id available = 0;
-  size_t batch_size;
+
+  /// Stores currently available IDs.
+  std::vector<id_generator> id_generators;
+
+  /// Returns the number of currently available IDs.
+  int32_t available_ids() const noexcept;
+
+  /// Returns the next available IDs.
+  /// @pre `available_ids() > 0`
+  id next_id();
+
+  /// Stores how many events inbound paths can still send us.
+  int32_t in_flight_events = 0;
```

Comment:
Why is this one signed? Do values < 0 make sense?

Reply:
Unsigned types are best avoided. The question is: do you ever wish to do arithmetic on a value? If the answer is yes than you should never use an unsigned type. To paraphrase Chandler Carruth: if you need more bits - get more bits! So the question I ask here is: are 2^31 possible values enough? Yes!

To add a bit more background: The reason why unsigned types are bad is that they force the compiler to emit outrageously expensive code. The standard guarantees that arithmetic on unsigned types overflows. For example, adding +2 to an `unsigned char` with current value `255` *must* produce `1`. This means the compiler *has to* perform bound checks *every single time*. The only exception here is `size_t`, because it automatically overflows since it uses the native word length. We can't escape `size_t` since it's used ubiquitously in the STL, but the only valid use case for unsigned types is for storing bit flags.

That's my reasoning behind not using any unsigned type other than `size_t`, which follows the recommendation of Stroustrup, Carruth, and others.

Reply:
Very interesting, I learned something new. So your approach is performance-based, while I was going with domain semantics. 

As a guideline, why don't we always use `size_t` for counters then? It seems that this type doesn't suffer from the performance problems but also provides a sensible value domain by default.